tests/query_params_test.go

```diff
@@ -0,0 +1,216 @@
+func isBadReq(resp []byte) bool {
+	if bytes.HasPrefix(resp, []byte(statusBadReq)) {
+		return true
+	}
+	return false
+}
+
+// Ensures Query() only returns a map of size one and verifies whether sending a
+// request with two values for the same key returns a []string of length 2
+// containing the correct values
+func TestMultipleQueryParametersSameKey(t *testing.T) {
+	var (
+		valueOne = "potatO"
+		valueTwo = "Tomato"
+		req      = []byte("GET /?vegetable=" + valueOne + "&vegetable=" + valueTwo + " HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	)
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		queryParams := req.URL.Query()
+		if len(queryParams) != 1 {
+			t.Errorf("len(queryParams): got %d, want %d", len(queryParams), 1)
+		}
+
+		vegetableParamValues := queryParams["vegetable"]
+		if len(vegetableParamValues) != 2 {
+			t.Errorf("len(vegetableQueryParams): got %d, want %d", len(vegetableParamValues), 2)
+		}
+		if vegetableParamValues[0] != valueOne || vegetableParamValues[1] != valueTwo {
+			t.Errorf("queryParams values: expected "+valueOne+" and "+valueTwo+", got %s and %s", vegetableParamValues[0], vegetableParamValues[1])
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+
+}
+
+func TestQueryParametersSameKeyDifferentCasing(t *testing.T) {
+
+	req := []byte("GET /?vegetable=potato&Vegetable=tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		queryParams := req.URL.Query()
+
+		if len(queryParams) != 2 {
+			t.Errorf("len(queryParams): got %d, want %d", len(queryParams), 2)
+		}
+
+		if len(queryParams["vegetable"]) != 1 || len(queryParams["Vegetable"]) != 1 {
+			t.Errorf("Expected one value for query parameter vegetables and one for Vegetables, got %d and %d", len(queryParams["vegetable"]), len(queryParams["Vegetable"]))
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+}
+
+// TestQueryParametersValidUnicode ensures keys and values that contain non-ASCII characters are parsed correctly
+func TestQueryParametersValidUnicode(t *testing.T) {
+	value := "ăȚâȘî"
+	req := []byte("GET /?vegetable=" + value + " HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		if valueReceived := req.URL.Query()["vegetable"][0]; valueReceived != value {
+			t.Errorf("queryParams values: got %s, want %s", valueReceived, value)
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
```

Comment:
I think this test (and the ones above) can be folded into a table-test.

Maybe a structure like this:

```
tests := []struct{
  name string
  req string
  want url.Values
}{{
  name: "same key",
  req: "/?vegetable=tomato&vegetable=potato",
  want: url.Values{
    "vegetable": []string{"tomato", "potato"},
  },
},{
  ...
}}

for _, test := range tests {
  t.Run(test.name, func(t *testing.T) {
     // similar to the existing tests
  })
```

https://blog.golang.org/subtests

Reply:
As far as I understand, table-tests give multiple inputs and outputs, but the test applied to them has the same logic. Here, the checks are slightly different, that's why I opted against. But I might be wrong.

Reply:
I refactored the code to include table-driven trests as much as possible in 12a32a